tests/integration/test_with_portainer_service.py

```diff
@@ -21,9 +88,265 @@ async def test_portainer_connection(
+    )
+    # Note:
+    # Alternatively, we could also check the sha of the contianer and assess the container is re-deployed
+    # But this takes time ti take affect and would require sleeps or retr ying polycies. So we dont do it for now. The following call can be used for this purpose:
+    # rawContainerImageAfter = _run_cmd("docker inspect $(docker service ps $(docker service ls | grep sleeper | cut -d ' ' -f1) | grep Running | cut -d ' ' -f1) | jq '.[0].Spec.ContainerSpec.Image'")
+
+    containerImageSHAAfter = rawContainerImageAfter.split("@")[1]
+    assert containerImageSHABefore != containerImageSHAAfter
+
+    ## Cleanup for subsequent tests: ...
+    await portainer.delete_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_bearer_code,
+        stack_id=stack_id,
+        endpoint_id=portainer_endpoint_id,
+    )
+
+
+async def test_portainer_raises_when_stack_already_present_and_can_delete(
+    event_loop: asyncio.AbstractEventLoop,
+    portainer_container: tuple[URL, str],
+    aiohttp_client_session: ClientSession,
+    portainer_bearer_code: str,
+    portainer_endpoint_id: int,
+    valid_docker_stack: ComposeSpecsDict,
+    docker_swarm: None,
+):
+    portainer_url, _ = portainer_container
+    swarm_id = await portainer.get_swarm_id(
+        portainer_url,
+        aiohttp_client_session,
+        portainer_bearer_code,
+        portainer_endpoint_id,
+    )
+    # Assuring a clean state by deleting any remnants
+    current_stack_name = "pytestintegration"
+    os.system("docker stack rm " + current_stack_name)
+
+    new_stack = await portainer.post_new_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_bearer_code,
+        swarm_id=swarm_id,
+        endpoint_id=portainer_endpoint_id,
+        stack_name=current_stack_name,
+        stack_cfg=valid_docker_stack,
+    )
+
+    async for attempt in AsyncRetrying(
+        stop=stop_after_attempt(10),
+        wait=wait_fixed(1),
+    ):
+        with attempt:
+            await portainer.get_current_stack_id(
+                base_url=portainer_url,
+                app_session=aiohttp_client_session,
+                bearer_code=portainer_bearer_code,
+                stack_name=current_stack_name,
+            )
+
+    with pytest.raises(exceptions.AutoDeployAgentException):
+        new_stack = await portainer.post_new_stack(
+            base_url=portainer_url,
+            app_session=aiohttp_client_session,
+            bearer_code=portainer_bearer_code,
+            swarm_id=swarm_id,
+            endpoint_id=portainer_endpoint_id,
+            stack_name=current_stack_name,
+            stack_cfg=valid_docker_stack,
+        )
+    ## Cleanup for subsequent tests: ...
+    stack_id = await portainer.get_current_stack_id(
```

Comment:
since you have the clean_stack fixture you should not need that part of the code.